### PR TITLE
Fix Response Without Cert

### LINKED ISSUE #374: Not able to get latest refactored code in libsaml.js file
Receiving error 'NO_SELECTED_CERTIFICATE' in deploy logs,
I have old libsaml.js , I saw lots of refactored code in libsaml.js file of this repo.
On 'npm install' library should get updated , but that is not doing any changes in that file.
Can you please look into this matter , it's urgent 

**Here is the code snippet which I have in my workspace -**
 `selection.forEach(function (signatureNode) {
                sig.signatureAlgorithm = opts.signatureAlgorithm;
                if (!opts.keyFile && !opts.metadata) {
                    throw new Error('ERR_UNDEFINED_SIGNATURE_VERIFIER_OPTIONS');
                }
                if (opts.keyFile) {
                    sig.keyInfoProvider = new xml_crypto_1.FileKeyInfo(opts.keyFile);
                }
                if (opts.metadata) {
                    var certificateNode = xpath_1.select(".//*[local-name(.)='X509Certificate']", signatureNode);
                    // certificate in metadata
                    var metadataCert = opts.metadata.getX509Certificate(certUse.signing);
                    // flattens the nested array of Certificates from each KeyDescriptor
                    if (Array.isArray(metadataCert)) {
                        metadataCert = utility_1.flattenDeep(metadataCert);
                    }
                    else if (typeof metadataCert === 'string') {
                        metadataCert = [metadataCert];
                    }
                    // normalise the certificate string
                    metadataCert = metadataCert.map(utility_1.default.normalizeCerString);
                    if (certificateNode.length === 0) {
                        throw new Error('NO_SELECTED_CERTIFICATE');
                    }`

Thanks . . !

### PR DESCRIPTION
**Pull request by natevecc**
Needed this for a client setup. Current implementation seems to only work if there is a certificate in the response. This changes it to fallback to use the first certificate in the metadata if none are provided via the response. Would appreciate some guidance in setting up tests for this.

Fixes #374

### CODE REVIEW

**Comment by tngan** [comment]
@natevecc According to the specification, SAML response doesn't have to include the certificate though. That's why we fallback to use the first certificate from the given metadata.

**Comment by natevecc** [comment]
@tngan right but the code wasn't doing that. It was erroring out saying it couldn't find a certificate even though it was included in the metadata. This change will now only error out if it can't find a certificate in the response or the metadata. Then it tries to validate the certificate in the response against the one in the metadata. If there is no certificate in the response it just uses the first certificate from the metadata.

I'll try and add some tests that illustrate these use cases soon.

**Comment by eqqnqqn** [comment]
Hello Tony,

I would like to add a new point before you rollout the correction.
As you said it, certificate is not mandatory in request and thanks to your correction, samlify will fallback to the certificate in metadata.

However, in my use case, I don’t trust the certificate in request so 
•	I would like to use exclusively the certificate in metadata 
•	and so generate error if in case of signature verification is needed, only request certificate is available (if during registration for example of a client, I forgot to fill its certificate in metadata)
 
So it could be useful to have new parameter(s) in server configuration to define 
* which certificate has the highest priority if both are present (in metadata and in request)
* and potentially if the use of this certificate is exclusive ( only this certificate is allowed (ie metadata) ) 
 
Thanks for your feedback
 
Best regards.

**Comment by KaniZ0r** [comment]
Hi @tngan 

You mentioned [here](https://github.com/tngan/samlify/issues/382#issuecomment-675874273) that you'll look in to this shortly.
Can you if you've plans to accept this PR and tag a release with it included?

Thanks in advance

**Comment by eqqnqqn** [comment]
Hello Tony, @tngan

Could you take a look to my comment, I would like to use exclusively the certificate in metadata, and this need is essential for our project. Could you include this point in the next release?
Thanks in advance.

> 
> 
> Hello Tony,
> 
> I would like to add a new point before you rollout the correction.
> As you said it, certificate is not mandatory in request and thanks to your correction, samlify will fallback to the certificate in metadata.
> 
> However, in my use case, I don’t trust the certificate in request so
> • I would like to use exclusively the certificate in metadata
> • and so generate error if in case of signature verification is needed, only request certificate is available (if during registration for example of a client, I forgot to fill its certificate in metadata)
> 
> So it could be useful to have new parameter(s) in server configuration to define
> 
>     * which certificate has the highest priority if both are present (in metadata and in request)
> 
>     * and potentially if the use of this certificate is exclusive ( only this certificate is allowed (ie metadata) )
> 
> 
> Thanks for your feedback. @tngan
> 
> Best regards.



Hello Tony, @tngan

Could you take a look to my comment please? I would like to use exclusively the certificate in metadata, instead of the one in request. This need is essential for our project. Could you accept this point and include it in the next release?

Thank you in advance.

**Comment by natevecc** [comment]
@tngan sorry I've been distracted from this and was wondering if you could let me know what you'd like done so this could get merged.

File: src/libsaml.ts [hunk 1]
```diff
@@ -366,11 +366,12 @@ const libSaml = () => {
           // normalise the certificate string
           metadataCert = metadataCert.map(utility.normalizeCerString);
 
-          if (certificateNode.length === 0) {
+          // no certificate in response or metadata
+          if (metadataCert.length === 0 && metadataCert.length === 0) {
```
Comment (ryanau) [inline src/libsaml.ts:370]: Is this supposed to be `certificateNode.length`?

**Comment by ryanau** [comment]
Any update on this PR? :D

**Comment by azuruce** [comment]
any update?

**Comment by KaniZ0r** [comment]
It seems that this is fixed in this commit: https://github.com/tngan/samlify/commit/49e25153dc2584461613ded9dfaf30f55e964ce4
I think this pull request can be closed now. @natevecc

**Comment by natevecc** [comment]
I'm closing this because as of version 2.7.7 the issue I had is no longer occurring.

**Pull request closed without merging**